File: ios/Packages/HapiKit/Sources/HapiProtocol/Models/FilesApi.swift
import Foundation

// Wire types of the git & files surface (A-M4a):
// `docs/api/client-contract/rest.md` "Git & files"; shapes from
// `shared/src/apiTypes.ts` (`CommandResponse`, `FileReadResponse`,
// `FileSearchResponse`, `ListDirectoryResponse`), field-for-field in lockstep
// with the Android port (`app.hapi.protocol.wire.FilesApi`). All are
// RPC-wrapped — HTTP 200 with `success: false` is a normal failure, callers
// must check. The session `DirectoryEntry` row already lives in
// `ApiResponses.swift` (shared with the machine list-directory RPC family).

/// `GET /api/sessions/:id/git-status|git-diff-numstat|git-diff-file` —
/// `CommandResponse`: the hub relays raw git stdout verbatim; clients parse
/// it themselves (`HapiProtocol/Git/` twins `web/src/lib/gitParsers.ts`).
public struct GitCommandResponse: Codable, Equatable, Sendable {
    public var success: Bool
    public var stdout: String?
    public var stderr: String?
    public var exitCode: Int?
    public var error: String?

    public init(
        success: Bool,
        stdout: String? = nil,
        stderr: String? = nil,
        exitCode: Int? = nil,
        error: String? = nil
    ) {
        self.success = success
        self.stdout = stdout
        self.stderr = stderr
        self.exitCode = exitCode
        self.error = error
    }
}

/// `GET /api/sessions/:id/file?path=` — `content` is **base64** (decode
/// before display).
public struct FileReadResponse: Codable, Equatable, Sendable {
    public var success: Bool
    public var content: String?
    /// Bytes on disk.
    public var size: Int?
    /// Epoch ms.
    public var modified: Double?
    public var error: String?

    public init(
        success: Bool,
        content: String? = nil,
        size: Int? = nil,
        modified: Double? = nil,
        error: String? = nil
    ) {
        self.success = success
        self.content = content
        self.size = size
        self.modified = modified
        self.error = error
    }
}

/// `GET /api/sessions/:id/files?query=&limit=` (ripgrep-backed search).
public struct FileSearchResponse: Codable, Equatable, Sendable {
    public var success: Bool
    public var files: [FileSearchItem]?
    public var error: String?

    public init(success: Bool, files: [FileSearchItem]? = nil, error: String? = nil) {
        self.success = success
        self.files = files
        self.error = error
    }
}

public struct FileSearchItem: Codable, Equatable, Sendable {
    public var fileName: String
    /// Directory part; empty at the session root.
    public var filePath: String
    public var fullPath: String
    /// `'file' | 'folder'` (search results are files in practice).
    public var fileType: String
    public var size: Int?
    /// Epoch ms.
    public var modified: Double?

    public init(
        fileName: String,
        filePath: String,
        fullPath: String,
        fileType: String,
        size: Int? = nil,
        modified: Double? = nil
    ) {
        self.fileName = fileName
        self.filePath = filePath
        self.fullPath = fullPath
        self.fileType = fileType
        self.size = size
        self.modified = modified
    }
}

/// `GET /api/sessions/:id/directory?path=` — empty path = session root.
/// Entries are the shared session `DirectoryEntry` rows (`ApiResponses.swift`).
public struct ListDirectoryResponse: Codable, Equatable, Sendable {
    public var success: Bool
    public var entries: [DirectoryEntry]?
    public var error: String?

    public init(success: Bool, entries: [DirectoryEntry]? = nil, error: String? = nil) {
        self.success = success
        self.entries = entries
        self.error = error
    }
}

File: android/core/protocol/src/main/kotlin/app/hapi/protocol/wire/FilesApi.kt
package app.hapi.protocol.wire

import kotlinx.serialization.Serializable

// Wire types of the git & files surface (B-M4c):
// `docs/api/client-contract/rest.md` "Git & files"; shapes from
// `shared/src/apiTypes.ts` (`CommandResponse`, `FileReadResponse`,
// `FileSearchResponse`, `ListDirectoryResponse`). All are RPC-wrapped —
// HTTP 200 with `success: false` is a normal failure, callers must check.

/**
 * `GET /api/sessions/:id/git-status|git-diff-numstat|git-diff-file` —
 * `CommandResponse`: the hub relays raw git stdout verbatim; clients parse it
 * themselves (`app.hapi.protocol.git` twins `web/src/lib/gitParsers.ts`).
 */
@Serializable
data class GitCommandResponse(
    val success: Boolean,
    val stdout: String? = null,
    val stderr: String? = null,
    val exitCode: Int? = null,
    val error: String? = null,
)

/** `GET /api/sessions/:id/file?path=` — [content] is **base64** (decode before display). */
@Serializable
data class FileReadResponse(
    val success: Boolean,
    val content: String? = null,
    /** Bytes on disk. */
    val size: Long? = null,
    /** Epoch ms (fs mtime — may arrive fractional; see [LenientEpochMs]). */
    @Serializable(with = LenientEpochMs::class)
    val modified: Long? = null,
    val error: String? = null,
)

/** `GET /api/sessions/:id/files?query=&limit=` (ripgrep-backed search). */
@Serializable
data class FileSearchResponse(
    val success: Boolean,
    val files: List<FileSearchItem>? = null,
    val error: String? = null,
)

@Serializable
data class FileSearchItem(
    val fileName: String,
    /** Directory part; empty at the session root. */
    val filePath: String,
    val fullPath: String,
    /** `'file' | 'folder'` (search results are files in practice). */
    val fileType: String,
    val size: Long? = null,
    /** Epoch ms (fs mtime — may arrive fractional; see [LenientEpochMs]). */
    @Serializable(with = LenientEpochMs::class)
    val modified: Long? = null,
)

/** `GET /api/sessions/:id/directory?path=` — empty path = session root. */
@Serializable
data class ListDirectoryResponse(
    val success: Boolean,
    val entries: List<DirectoryEntry>? = null,
    val error: String? = null,
)

/**
 * `DirectoryEntry` (`shared/src/apiTypes.ts`). The machine-RPC twin with
 * `isGitRepo` is [MachineDirectoryEntry].
 */
@Serializable
data class DirectoryEntry(
    val name: String,
    /** `'file' | 'directory' | 'other'`. */
    val type: String,
    val size: Long? = null,
    /** Epoch ms (fs mtime — may arrive fractional; see [LenientEpochMs]). */
    @Serializable(with = LenientEpochMs::class)
    val modified: Long? = null,
)
